lightning-block-sync/src/lib.rs

```diff
@@ -0,0 +1,70 @@
+//! A lightweight client for keeping in sync with chain activity.
+//!
+//! Defines a [`BlockSource`] trait, which is an asynchronous interface for retrieving block headers
+//! and data.
+//!
+//! [`BlockSource`]: trait.BlockSource.html
+
+use bitcoin::blockdata::block::{Block, BlockHeader};
+use bitcoin::hash_types::BlockHash;
+use bitcoin::util::uint::Uint256;
+
+use std::future::Future;
+use std::pin::Pin;
+
+/// Abstract type for retrieving block headers and data.
+pub trait BlockSource : Sync + Send {
+	/// Returns the header for a given hash. A height hint may be provided in case a source cannot
+	/// easily find headers based on a hash. This is merely a hint and thus the returned header must
+	/// have the same hash as was requested. Otherwise, an error must be returned.
+	///
+	/// Implementations that cannot find headers based on the hash should return a `Transient` error
+	/// when `height_hint` is `None`. In such a case, `get_best_block` should never return `None`
+	/// for the height. Otherwise, the source could not be used independently for an initial sync.
+	fn get_header<'a>(&'a mut self, header_hash: &'a BlockHash, height_hint: Option<u32>) -> AsyncBlockSourceResult<'a, BlockHeaderData>;
+
+	/// Returns the block for a given hash. A headers-only block source should return a `Transient`
+	/// error.
+	fn get_block<'a>(&'a mut self, header_hash: &'a BlockHash) -> AsyncBlockSourceResult<'a, Block>;
+
+	/// Returns the hash of the best block and, optionally, its height. The height is passed to
+	/// `get_header` to allow a more efficient lookup on some block sources.
```

Comment:
The forthcoming `MicroSPVClient` would pass the height from either the result of calling `get_best_block` or by computing it from a `BlockHeaderData`.

I can write a better summary at the trait-level, but I think this raises an important point regarding the interface. Namely, in what case would a source not provide a height? If `get_best_block` returns a `None` height with the hash, how could a call to `get_header` on the same source return a `BlockHeaderData` for the hash given that it has a `height` field to populate?

@TheBlueMatt Should we drop `Option` from around the height in these two functions? Or at very least from `get_best_block`? I suppose we may want to call `get_header` without a hint if we deserialized a listener that only had the most recent block hash.

Reply:
Hmm, good question! I agree its maybe dubious to claim there will be an implementation that polls the best chain tip and only gets a hash, but its still somewhat nice to have - it wouldn't be unheard of, and requiring less data for the best block polling may leave some other things (like headers over radio) more easy to implement. I'd need to dig into the `MicroSPVClient`, but is it not the case that `height_hint` will *never* be None unless a source returned None for the height in `get_best_block` (or is it the case that you can hit it if one source returns None but another source returns Some?)? That would simplify the docs somewhat and I'm not sure it generates a lot of code in the `MicroSPVClient`?

Reply:
The multiple-source case is handled in `ChainMultiplexer`, which uses a set of `ChainPoller`s (one for each block source). So the result of `get_best_block` is never fed to `get_header` of a different source. So, yeah, I think the docs can be simplified a bit with this invariant in mind.

But note that relatedly, as it stands, `get_header` must return `BlockHeaderData` containing a height. If it were changed to be `Option`, this would prevent the fork detection logic from functioning. And more importantly, a height is eventually needed to pass to `block_connected` and `block_disconnected`.

IIUC, then, a headers-only source must return the height along with the header. Were you suggesting that it may be easier to implement such a source if this were not the case? We could conceivably make the height field an `Option`, but it would complicate the code quite a bit. Trying different source would be predicated on this field being `None` rather than on whether there was an error. I suppose that could be abstracted away to `ChainMultiplexer` if needed. Just want to make sure I understand the use case.